Sources/Core/Package.swift

```
@@ -0,0 +1,287 @@
+public enum PackageType: String {
+    case framework = "FMWK"
+    case bundle = "BNDL"
+    case dSYM = "dSYM"
+}
+
+public struct Package {
+    
+    // MARK: - Constants
+    
+    struct Constants {
+        static var lipoArchitecturesMessage: String = "Architectures in the fat file:"
+        static var lipoNonFatFileMessage: String = "Non-fat file:"
+    }
+    
+    // MARK: - Attributes
+    
+    private let path: Path
+    
+    // MARK: - Init
+    
+    public init(path: Path) {
+        self.path = path
+    }
+    
+    // MARK: - Public
+    
+    /// Returns the path of the binary inside the package.
+    ///
+    /// - Returns: path of the binary (if it exists).
+    public func binaryPath() -> Path? {
+        guard let bundle = Bundle(path: path.string) else { return nil }
+        guard let packageType = packageType() else { return nil }
+        switch packageType {
+        case .framework, .bundle:
+            return path + (bundle.object(forInfoDictionaryKey: "CFBundleExecutable") as? String).map({Path($0)})!
+        case .dSYM:
+            let binaryName = path.url.deletingPathExtension().deletingPathExtension().lastPathComponent
+            if !binaryName.isEmpty {
+                return path + Path("Contents/Resources/DWARF/\(binaryName)")
+            }
+        }
+        return nil
+    }
+    
+    /// Returns the package type.
+    ///
+    /// - Returns: package type.
+    public func packageType() -> PackageType? {
+        guard let bundle = Bundle(path: path.string) else { return nil }
+        guard let bundlePackageType = bundle.object(forInfoDictionaryKey: "CFBundlePackageType") as? String else { return nil }
+        return PackageType(rawValue: bundlePackageType)
+    }
+    
+    /// Returns the supported architectures of the given package.
+    ///
+    /// - Returns: all the supported architectures.
+    public func architectures() -> [String] {
+        guard let _binaryPath = binaryPath() else { return [] }
```

It is ok that the function returns an empty array in case of nullable binary path?

---

We could throw an error when we detect a malformed package but we'd have to update all the methods to throw. Do you think that should be handled throwing or leave it up to the developers to handle it from outside?

---

I thought about throwing, but I'm not sure that it's the best approach.

---

I'm not sure either, what do you think @esttorhe?